onnxruntime/core/framework/utils.cc

```diff
@@ -629,16 +624,38 @@ void DumpNodeOutputs(OpKernelContext& context, const Node& node, const SessionSt
           if (DEBUG_NODE_INPUTS_OUTPUTS > 1) {
             // check tensor is on CPU before dumping it
             auto& tensor_location = tensor.Location();
-            auto* provider = execution_providers.Get(tensor_location);
-            if (!provider) {
-              provider = cpu_execution_provider;
-            }
-
-            if (provider == cpu_execution_provider || tensor_location.mem_type == OrtMemTypeCPUOutput) {
-              const auto data_type = tensor.DataType();
+            const auto data_type = tensor.DataType();
+            if (tensor_location.device.Type() == OrtDevice::CPU || tensor_location.mem_type == OrtMemTypeCPUOutput) {
               DispatchOnTensorType(data_type, DumpTensor, tensor, shape);
             } else {
-              std::cout << " is not on CPU. Provider=" << provider->Type() << "\n";
+              std::cout << tensor_location << "\n";
+
+#ifdef USE_CUDA
+              // Dumping GPU only when cuda is enabled. Most op has only one output, so put GPU related code here to get best performance.
+              const auto& execution_providers = session_state.GetExecutionProviders();
+              const auto* cuda_execution_provider = execution_providers.Get(onnxruntime::kCudaExecutionProvider);
+              if (cuda_execution_provider == nullptr) {
+                continue;
+              }
+
+              auto gpu_data_transfer = cuda_execution_provider->GetDataTransfer();
```

Comment:
I don't think this should be calling GetDataTransfer as that creates a new instance each time. There's a data transfer manager in session_state that probably should be used instead.

Reply:
updated to use data transfer manager.